pandas/tests/frame/test_indexing.py

```diff
@@ -2590,3 +2590,32 @@ def test_head_tail(self):
         empty_df = DataFrame()
         assert_frame_equal(empty_df.tail(), empty_df)
         assert_frame_equal(empty_df.head(), empty_df)
+
+    def test_type_error_multiindex(self):
+        # See gh-12218
+        df = DataFrame(columns=['i', 'c', 'x', 'y'],
+                       data=[[0, 0, 1, 2], [1, 0, 3, 4],
+                             [0, 1, 1, 2], [1, 1, 3, 4]])
+        dg = df.pivot_table(index='i', columns='c',
+                            values=['x', 'y'])
+
+        with assertRaisesRegexp(TypeError, "is an invalid key"):
+            str(dg[:, 0])
+
+        columns = MultiIndex(levels=[['x', 'y'], [0, 1]],
+                             labels=[[0, 1], [0, 0]],
+                             names=[None, 'c'])
+        values = np.array([[1, 2], [3, 4]], dtype=np.int64)
+        index = Index(data=[0, 1], dtype='int64', name='i')
```

Comment:
oh I see you the index has a name, ok then.

Still construct like this 

``index = Index(range(2), name='i')`` it automatically is int64
you can just construct 
```
expected = DataFrame([[1, 2], [3, 4], columns=columns, dtype='int64')
```